File: fabric-audit-agent/core/whatif.js
import { DEFAULT_CONFIG } from './config.js';

/**
 * Project the capacity impact of a proposed asset against current facts. Pure.
 * @param {object} facts  current estate facts (reads facts.capacity)
 * @param {{ kind?:string, sizeGB?:number, refreshAt?:string, name?:string }} proposed
 * @param {object} [config]
 * @returns {{ proposed:object, impacts:string[], riskScore:number, verdict:'safe'|'risky'|'blocked' }}
 */
export function assessWhatIf(facts = {}, proposed = {}, config = DEFAULT_CONFIG) {
  const c = facts?.capacity ?? {};
  const impacts = [];
  let risk = 0;

  if (proposed.refreshAt) {
    const sameWindow = (c.refreshes ?? []).filter(r => r.scheduledAt === proposed.refreshAt);
    if (sameWindow.length >= 1) {
      impacts.push(`Refreshing at ${proposed.refreshAt} joins ${sameWindow.length} existing refresh(es) — worsens contention.`);
      risk += (sameWindow.length >= (config.capacity.contentionMin - 1)) ? 2 : 1;
    }
  }
  if (proposed.kind === 'model' && (proposed.sizeGB ?? 0) >= config.capacity.oversizedGB) {
    impacts.push(`Proposed model is ${proposed.sizeGB} GB (>= ${config.capacity.oversizedGB} GB oversized threshold).`);
    risk += 1;
  }
  if ((c.peakCuPct ?? 0) >= config.capacity.throttleWarnPct) {
    impacts.push(`Capacity ${c.capacityId ?? ''} already peaks at ${c.peakCuPct}% CU — little headroom for new load.`.trim());
    risk += 2;
  }

  const verdict = risk >= 4 ? 'blocked' : risk >= 2 ? 'risky' : 'safe';
  return { proposed, impacts, riskScore: risk, verdict };
}

File: fabric-audit-agent-app/fabric_audit_agent/whatif.py
"""Project the capacity impact of a proposed asset. Port of ``core/whatif.js``. Pure."""
from .config import DEFAULT_CONFIG


def assess_what_if(facts=None, proposed=None, config=None):
    facts = facts or {}
    proposed = proposed or {}
    config = config or DEFAULT_CONFIG
    c = facts.get("capacity") or {}
    cap = config["capacity"]
    impacts = []
    risk = 0

    if proposed.get("refreshAt"):
        same_window = [r for r in (c.get("refreshes") or []) if r.get("scheduledAt") == proposed["refreshAt"]]
        if len(same_window) >= 1:
            impacts.append(f"Refreshing at {proposed['refreshAt']} joins {len(same_window)} existing refresh(es) — worsens contention.")
            risk += 2 if len(same_window) >= (cap["contentionMin"] - 1) else 1
    if proposed.get("kind") == "model" and (proposed.get("sizeGB") or 0) >= cap["oversizedGB"]:
        impacts.append(f"Proposed model is {proposed.get('sizeGB')} GB (>= {cap['oversizedGB']} GB oversized threshold).")
        risk += 1
    if (c.get("peakCuPct") or 0) >= cap["throttleWarnPct"]:
        impacts.append(f"Capacity {c.get('capacityId') or ''} already peaks at {c.get('peakCuPct')}% CU — little headroom for new load.".strip())
        risk += 2

    verdict = "blocked" if risk >= 4 else ("risky" if risk >= 2 else "safe")
    return {"proposed": proposed, "impacts": impacts, "riskScore": risk, "verdict": verdict}
